File: Data Structures & Algorithms/permutations/submission-0.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        resultList = []

        def build(numsSubset, result):
            if len(numsSubset) == 0:
                resultList.append(result.copy())
            for i in range(len(numsSubset)):
                val = numsSubset.pop(i)
                result.append(val)
                build(numsSubset, result)
                result.pop()
                numsSubset.insert(i, val)

        build(nums, [])
        return resultList

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-0.java
class Solution {
    public int search(int[] nums, int t) {
        int result = -1;
        int l = 0;
        int r = nums.length;
        int lastNum = nums[nums.length - 1];
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (t == nums[mid]) {
                result = mid;
                break;
            }
            if (t > lastNum && nums[mid] <= lastNum) {
                r = mid;
            }
            else if (t <= lastNum && nums[mid] > lastNum) {
                l = mid + 1;
            }
            else {
                if (t < nums[mid]) {
                    r = mid;
                }
                else {
                    l = mid + 1;
                }
            }
        }
        return result;
    }
}
